File: client/crypto-sdk-android/crypto-sdk/src/main/java/com/orbs/client/OrbsContract.java
package com.orbs.client;


import com.google.gson.Gson;
import com.orbs.cryptosdk.Address;

public class OrbsContract {
  private final OrbsClient orbsClient;
  private final Address contractAddress;

  public OrbsContract(OrbsClient client, String contractName) throws Exception {
    this.orbsClient = client;
    String contractHash = OrbsHashUtils.bytesToHex(OrbsHashUtils.hash256(contractName));
    this.contractAddress = new Address(contractHash, client.senderAddress.virtualChainId, client.senderAddress.networkId);
  }

  public SendTransactionResponse sendTransaction(String methodName, Object[] args) throws Exception {
    String payload = generateSendTransactionPayload(methodName, args);
    return orbsClient.sendTransaction(this.contractAddress, payload);
  }

  public String call(String methodName, Object[] args) throws Exception {
    String payload = generateCallPayload(methodName, args);
    return orbsClient.call(this.contractAddress, payload);
  }

  public String generateSendTransactionPayload(String methodName, Object[] args) throws Exception {
    SendTransactionPayload payload = new SendTransactionPayload.Builder()
            .withMethod(methodName)
            .withArgs(args)
            .build();
    Gson gson = new Gson();
    return gson.toJson(payload);
  }

  public String generateCallPayload(String methodName, Object[] args) {
    CallPayload payload = new CallPayload();
    payload.method = methodName;
    payload.args = args;
    Gson gson = new Gson();
    return gson.toJson(payload);
  }

  public OrbsClient getOrbsClient() {
    return orbsClient;
  }

  public Address getContractAddress() {
    return contractAddress;
  }
}

File: client/client-sdk-javascript/src/orbs-contract.ts
import { OrbsClient } from "./orbs-client";
import { Address } from "./address";
import * as crypto from "crypto";

export type OrbsContractMethodArgs = [string | number] | any[];

export class OrbsContract {
  public orbsClient: OrbsClient;
  public contractAddress: Address;

  constructor(orbsClient: OrbsClient, contractName: string) {
    this.orbsClient = orbsClient;
    const contractKey = crypto.createHash("sha256").update(contractName).digest("hex");
    this.contractAddress = new Address(
      contractKey,
      this.orbsClient.senderAddress.virtualChainId,
      this.orbsClient.senderAddress.networkId.toString()
    );
  }

  public async sendTransaction(methodName: string, args: OrbsContractMethodArgs) {
    const payload = this.generateSendTransactionPayload(methodName, args);
    return await this.orbsClient.sendTransaction(this.contractAddress, payload);
  }

  public generateSendTransactionPayload(methodName: string, args: OrbsContractMethodArgs) {
    let argsToUse = args;
    if (args == undefined) {
      argsToUse = [];
    }
    return JSON.stringify({
      method: methodName,
      args: argsToUse
    });
  }

  public async call(methodName: string, args: OrbsContractMethodArgs) {
    const payload = this.generateCallPayload(methodName, args);
    return this.orbsClient.call(this.contractAddress, payload);
  }

  public generateCallPayload(methodName: string, args: OrbsContractMethodArgs) {
    let argsToUse = args;
    if (args == undefined) {
      argsToUse = [];
    }
    return JSON.stringify({
      method: methodName,
      args: argsToUse
    });
  }
}
